src/frontend/src/optimizer/plan_node/batch_hash_join.rs

```diff
@@ -72,13 +72,16 @@ impl BatchHashJoin {
     ) -> Distribution {
         match (left, right) {
             (Distribution::Single, Distribution::Single) => Distribution::Single,
-            (Distribution::HashShard(_), Distribution::HashShard(_) | Distribution::SomeShard) => {
-                l2o_mapping.rewrite_provided_distribution(left)
-            }
-            (Distribution::SomeShard, Distribution::HashShard(_)) => {
+            (
+                Distribution::HashShard(_),
+                Distribution::HashShard(_) | Distribution::UpstreamHashShard(_),
+            ) => l2o_mapping.rewrite_provided_distribution(left),
+            (Distribution::UpstreamHashShard(_), Distribution::HashShard(_)) => {
                 r2o_mapping.rewrite_provided_distribution(right)
             }
-            (Distribution::SomeShard, Distribution::SomeShard) => Distribution::SomeShard,
+            (Distribution::UpstreamHashShard(_), Distribution::UpstreamHashShard(_)) => {
+                Distribution::SomeShard
+            }
```

Comment:
I think if `UpstreamHashShard` appears on one side of the hash join's input, it will be a wrong hash join plan which can not make sure the sharding of its two sides is the same.

Reply:
Oh this is used after enforce distribution? I misunderstood it.